### PR TITLE
Add waitWriteFinish option

### PR DESCRIPTION
**Pull request by alexpusch**
The `add` event will fire when a file first appear on disk, before the entire file has been written. Furthermore, in some cases some `change` events will be emitted while the file is being written.
In some cases, especially when watching for large files there will be a need to wait for the write operation to finish before responding to the file creation.

Current event throttling is insufficient to handle this case. For a large enough file, or slow enough write, change event might be emitted for several seconds after the file have appeared.

Setting `waitWriteFinish` to `true` will poll a newly created file size, holding its `add` and `change` events until its size will not change for `writeFinishThreshold` amount of time.

**Note**
Some tests failed to pass for me. I've patched them in the first commit, but I'm not confident in the fix. Please advise.
Fix might be related to issue #273 or issue #258

### CODE REVIEW

**Comment by es128** [comment]
Thanks for the PR. Looks like a useful option, and I think I'm alright with the solution concept.

I'll need to take some time to review carefully and try the functionality out before merging. 

And we do need to get your new tests passing on Travis, which [is not consistently the case now](https://travis-ci.org/paulmillr/chokidar/builds/80001124). It might have to do with how the synthetic delays are laid out, which can be really tricky to get right. I'd suggest adding more delay wrappers until the tests pass consistently, and then start removing ones that seem extraneous a little at a time watching the results with each small change.

**Comment by alexpusch** [comment]
I've now passed my tests. There are some unrelated tests that fail. The failed tests actually helped me find an edge case that I didn't handle. You can see it in this [commit](https://github.com/alexpusch/chokidar/commit/62f948181d92f9b60278f213207cbb1a4a59b359). I'll rebase this commit to the main one after you review the code.

**Comment by es128** [comment]
@alexpusch please rebase. Would also appreciate if you could take a pass at matching code style better - notably spacing in your `if` blocks and semicolons.

**Comment by alexpusch** [comment]
I went over the style issues and updated with 1.0.6. Rebased all fixes into current commit

File: index.js [hunk 1]
```diff
@@ -368,6 +434,12 @@ FSWatcher.prototype._remove = function(directory, item) {
   var wasTracked = parent.has(item);
   parent.remove(item);
 
+  // If we wait for this file to be fully written, cancel the wait.
+  if(this.waitWriteFinish && this._pendingWrites[path]){
```
Comment (paulmillr) [inline index.js]: space

Comment (paulmillr) [inline index.js]: why `this.waitWriteFinish` instead of `options.waitWriteFinish`

Comment (alexpusch) [inline index.js]: Rebased fix

**Comment by es128** [comment]
@alexpusch I haven't found the time quite yet to play with this locally, but hope to get to it shortly.

In the meantime, I'm wondering why 2000 is an appropriate default - can't it be any lower?

The latest commit also reintroduces some code style issues with a bunch of `if(`s missing a space.

File: README.md [hunk 2]
```diff
@@ -167,6 +167,20 @@ subdirectories will be traversed.
...
   ([see list of binary extensions](https://github.com/sindresorhus/binary-extensions/blob/master/binary-extensions.json))
 
+#### Waiting for write operation to finish
+* `waitWriteFinish` (default: `false`).
+The `add` event will fire when a file first appear on disk, before the entire
+file has been written. Furthermore, in some cases some `change` events will be emitted while the file is being written.
+In some cases, especially when watching for large files there will be a need to 
+wait for the write operation to finish before responding to the file creation.
+Setting `waitWriteFinish` to `true` will poll a newly created file size, holding 
+its `add` and `change` events until its size will not change for a configurable amount of time.
```
Comment (es128) [inline README.md]: until the size does not change for a configurable amount of time

Comment (es128) [inline README.md]: The "Use with caution." sentence should be moved up here.

File: README.md [hunk 3]
```diff
@@ -167,6 +167,20 @@ subdirectories will be traversed.
...
+* `writeFinishThreshold` (default: 2000).
+Amount of time in milliseconds for a file size to remain constant before emitting its 
+`add` event. Unfortunately this duration is heavily dependent on OS and local hardware.
```
Comment (es128) [inline README.md]: Don't use the word "unfortunately"

The appropriate duration setting is heavily dependent on the OS and hardware.

File: index.js [hunk 4]
```diff
@@ -194,6 +215,51 @@ FSWatcher.prototype._throttle = function(action, path, timeout) {
...
+      if(this._pendingWrites[path] === undefined){
+        this._pendingWrites[path] = {
+          creationTime: now,
+          cancelWait: function(){
+            delete this._pendingWrites[path];
+            clearTimeout(timeoutHandler);
+            return callback();
+          }.bind(this)
+        }
+        return timeoutHandler = setTimeout(awaitWriteFinish.bind(this, curStat), this.options.interval);
```
Comment (es128) [inline index.js]: Conflating the use of the `interval` option here, which is intended (and documented) as applying to the `usePolling` option. This probably needs a dedicated property for `awaitWriteFinishInterval`.

Comment (es128) [inline index.js]: Maybe the options need to be laid out more like

```js
awaitWriteFinish: {
    enabled: true,
    pollInterval: 250,
    stabilityThreshold: 2000
}
```

**Comment by alexpusch** [comment]
The number 2000 came of some local testing I've done. I've noticed that simple cp or mv commands tend to have 'pauses' of a second or two during the write process. It has a lot to do with the speed/number of disks you have, RAID setting, etc. I'll research it some more.

As for the option structure - Currently there is no other option that have its options encapsulated in a dedicated object. I can see how this approach tidying up the options object. Do you think it's worth going out of convention? 

Maybe the use of this option with the default values can be a simple
```js
awaitWriteFinish: true
```
And if the user wishes to change some parameters:
```js
awaitWriteFinish: {
   pollInterval: 250,
   stabilityThreshold: 2000
}
```

**Comment by es128** [comment]
:+1: I'm good with that suggestion for the config

As for the default values, if those are the values that work best for you, I'm ok with sticking to that for now and seeing what happens when we get feedback from other users of this feature.
